### DISCUSSION TITLE
Performance next steps

### DISCUSSION THREAD

**Discussion by iincer** [pacti-org/pacti#335] [General]
Pacti's webpage should have an entry for performance, as users would be interested to know this.
This brings the question of how we should be running the performance tests. So far, benchmarks have been generated from our computers. Can we automate benchmarking by generating results from github actions?
Also, regarding performance, we have learned much about possible simplifications in recent experiments run by @NicolasRouquette.
Finally, a likely high-impact change is to make use of Cython in the project.

**Comment by murrayrm** [comment]
For benchmarking, you might consider using [asv](https://github.com/airspeed-velocity/asv).

We've used Cython for other projects and find it to be a good way to get a lot of speedup.  Some benchmarks [here](https://www.biorxiv.org/content/10.1101/121152v2.full.pdf), summarized in this table:

<img width="641" alt="Screenshot 2023-09-04 at 12 33 50 AM" src="https://github.com/pacti-org/pacti/assets/293362/d701034b-1083-422a-9dc1-ccfaefe7f178">

**Reply by iincer** [reply]
Thanks, @murrayrm! I didn't know this library. It sounds like a useful way to track the impact of code changes on performance.
We have to figure out a strategy to make and keep track of these comparisons.

**Comment by NicolasRouquette** [comment]
In addition to the above suggestions, there is also an opportunity to refactor the internal representation to remove equality constraints as described in this issue: https://github.com/pacti-org/pacti/issues/344
